en/flying/terrain_following_holding.md

```diff
@@ -0,0 +1,81 @@
+As the distance to ground changes, the altitude setpoint adjusts to keep the height above ground constant.
+
+At higher altitudes (when the estimator reports that the distance sensor data is invalid) the vehicle switches to *altitude following*, and will typically fly at a near-constant height above mean sea level (AMSL) using the barometer for altitude data.
+
+> **Note** More precisely, the vehicle will use the *primary source of altitude data* as defined in [EKF2_HGT_MODE](../advanced_config/parameter_reference.md#EKF2_HGT_MODE).
+    This is, by default, the barometer.
+
+Terrain following is enabled by setting [MPC_ALT_MODE](../advanced_config/parameter_reference.md#MPC_ALT_MODE) to `1`.
+
+
+## Terrain Hold {#terrain_hold}
+
+Terrain holding uses a distance sensor to help a vehicle to better maintain a constant height above ground in [position](../flight_modes/position_mc.md) or [altitude](../flight_modes/altitude_mc.md) modes, when horizontally stationary at low altitude.
+This allows a vehicle to avoid altitude changes due to barometer drift or excessive barometer interference from rotor wash.
+
+> **Note** Terrain hold is enabled only in *multicopters* and *VTOL vehicles in MC mode*, and can be used in [position](../flight_modes/position_mc.md) and [altitude modes](../flight_modes/altitude_mc.md).
+
+When moving horizontally (`speed >` [MPC_HOLD_MAX_XY](../advanced_config/parameter_reference.md#MPC_HOLD_MAX_XY), or above the altitude where the distance sensor is providing valid data, the vehicle will instead use the *primary source of altitude data* ([EKF2_HGT_MODE](../advanced_config/parameter_reference.md#EKF2_HGT_MODE)).
+Typically this is the barometer.
+
+Terrain holding is enabled by setting [MPC_ALT_MODE](../advanced_config/parameter_reference.md#MPC_ALT_MODE) to `2`.
+
+> **Note** *Terrain hold* is implemented similarly to *terrain following.
+  It uses the output of the EKF estimator to provide the altitude estimate, and the estimated terrain altitude (calculated from distance sensor measurements using another estimator) to provide the altitude setpoint.
+  As the distance to ground changes due to external forces, the altitude setpoint adjusts to keep the height above ground constant.
+
+
+## Distance Sensor as Primary Source of Height {#distance_sensor_primary_altitude_source}
+
+PX4 allows you to make a distance sensor the *primary source of altitude data*.
+This may be useful when no barometer is available, or for applications when the vehicle is *guaranteed* to only fly over a near-flat surface (e.g. indoors).
+
+> **Tip** The default and preferred altitude sensor for most use cases is the barometer (when available).
+
+When using a distance sensor as the primary source of height, fliers should be aware:
+- Flying over obstacles can lead to the estimator rejecting data (due to internal data consistency checks).
+  This will result in TBD.
```

Comment:
@RomanBapst What happens if the estimator rejects the data?

Reply:
@hamishwillee the estimator will reject the data for a certain amount of time until the estimated variance (uncertainty) of that state exceeds a threshold. After that the estimator will perform a state reset which should make data consistent again. Think about it like "ok, my data is inconsistent, let me add an offset to the estimated altitude such that the range finder data is consistent with my altitude again. "

Reply:
@hamishwillee In practice the user could notice a drift in altititude (e.g. drone not maintaining altitude well) during the time when range data is rejected.

Reply:
@hamishwillee We are talking about a duration of roughly 6 seconds.

Reply:
@RomanBapst I'll rewrite this, but I'm really not 100% certain how time vs variance vs reset works

You say "for a certain amount of time until the estimated variance (uncertainty) of that state exceeds a threshold" and "We are talking about a duration of roughly 6 seconds.". Which makes me think that the threshhold/uncertainty is defined in terms of time? 

So consider this scenario were vehicle moving slowly up hill and maintaining altitude above ground. Accelerometer data will indicates a climb whereas the range finder data claims that the drone is not increasing altitude.
I think you are saying that:
- the estimator will *immediately* reject the rangefinder (?) data
- at this point the altitude starts to drift - ie it is hard to maintain constant altitude.
- after 6 or so seconds the system resets and the estimated altitude matches. 
- If the drone was to start moving over a level field the data would still be inconsistent right because of the existing offset - so there would still be a reset after 6 seconds right?

Reply:
PS Made minor change to this section, but did not capture all of above.

Reply:
@hamishwillee The reset is triggered based on time, see [here](https://github.com/PX4/ecl/blob/master/EKF/control.cpp#L738)
So basically if the range finder data is rejected for 5 seconds then a reset will be performed.
Whether range data is rejected or not depends on innovation consistency checks. Those checks take into account the error between measurement and current state as well as the estimated variance of the state and the variance of the measurement itself.
If the filter is very confident about the estimated state (low variance) then a measurement will be rejected "quicker".

Reply:
@hamishwillee 

> the estimator will immediately reject the rangefinder (?) data as soon as they differ

As explained above it depends on innovation consistency checks, it's not just a fixed value.
So no, it will not immediately reject the data.

> at this point the altitude starts to drift as rangefinder data is rejected and altitude is maintained based on the accelerometer - ie it is hard to maintain constant altitude

yes

> There is some threshhold for trusting accelerometer data only - about 6 seconds or so

5 seconds, see [here](https://github.com/PX4/ecl/blob/master/EKF/control.cpp#L738)

> As I see it the only way that the rangefinder can come consistent again is if the vehicle then goes down the hill by the same amount, within the 6 seconds

Yes, that would work

> Anyway, if they aren't consistent after that time, the estimator resets and effectively adds an offset to the rangefinder estimated altitude to make it match what the acc. thinks it is.

No, the estimator resets the state (in this case height) to the current distance sensor data.


> The altitude drifts during the 6 seconds where they don't match because the accelerometer estimate is going to be increasingly unreliable

Predicting the state reliably  using only the accelerometer cannot be done for long with the sensors we are using.